core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1986,101 +1965,125 @@ private[controller] class ControllerStats extends KafkaMetricsGroup {
 
 sealed trait ControllerEvent {
   def state: ControllerState
+  def preempt(): Unit
```

Comment:
I would add a documentation comment to this method explaining that this method will not be executed by the controller thread but instead it will be executed by some other thread.